File: server/Server/Models/Domain/WorkflowStage.cs
using Microsoft.AspNetCore.Identity;
using server.Models.Domain.Enums;

namespace server.Models.Domain;

public class WorkflowStage
{
    public int WorkflowStageId { get; set; }
    public string Name { get; set; } = string.Empty;
    public string? Description { get; set; }
    public int StageOrder { get; set; }
    public WorkflowStageType StageType { get; set; }
    public bool IsInitialStage { get; set; }
    public bool IsFinalStage { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }

    // Foreign Keys
    public int WorkflowId { get; set; }
    public int? InputDataSourceId { get; set; }
    public int? TargetDataSourceId { get; set; }

    // Navigation Properties
    public virtual Workflow Workflow { get; set; } = null!; // Required parent workflow
    public virtual DataSource? InputDataSource { get; set; }
    public virtual DataSource? TargetDataSource { get; set; }

    // Stage Pipeline Relationships via connections
    public virtual ICollection<WorkflowStageConnection> IncomingConnections { get; set; } = [];
    public virtual ICollection<WorkflowStageConnection> OutgoingConnections { get; set; } = [];

    // Tasks currently at this stage
    public virtual ICollection<Task> TasksAtThisStage { get; set; } = [];

    // User assignments with roles for this stage
    public virtual ICollection<WorkflowStageAssignment> StageAssignments { get; set; } = [];
}

File: frontend/src/types/workflow/workflowstage.ts
import type { ProjectMember } from "../projectMember";

export enum WorkflowStageType {
    ANNOTATION = "ANNOTATION",
    SUSPENDED = "SUSPENDED",
    DEFERRED = "DEFERRED",
    REVIEW = "REVIEW",
    REQUIRES_CHANGES = "REQUIRES_CHANGES",
    ACCEPTED = "ACCEPTED"
}

export interface WorkflowStage {
    id: number;
    name: string;
    description?: string;
    stageOrder: number;
    stageType?: WorkflowStageType;
    isInitialStage: boolean;
    isFinalStage: boolean;
    createdAt: string;
    updatedAt: string;
    workflowId: number;
    inputDataSourceId?: number;
    targetDataSourceId?: number;
    
    // Pipeline relationships
    incomingConnections: WorkflowStageConnection[];
    outgoingConnections: WorkflowStageConnection[];
    
    // User assignments
    assignments: WorkflowStageAssignment[];
}

export interface WorkflowStageConnection {
    id: number;
    fromStageId: number;
    toStageId: number;
    condition?: string;
    createdAt: string;
    updatedAt: string;
}

export interface WorkflowStageAssignment {
    id: number;
    workflowStageId: number;
    projectMember: ProjectMember;
    createdAt: string;
    updatedAt: string;
}

export interface WorkflowStagePipeline {
    id: number;
    name: string;
    description?: string;
    stageOrder: number;
    stageType?: WorkflowStageType;
    isInitialStage: boolean;
    isFinalStage: boolean;
    previousStageIds: number[];
    nextStageIds: number[];
    assignedUserCount: number;
}

export interface CreateWorkflowStageRequest {
    name: string;
    description?: string;
    stageOrder: number;
    stageType?: WorkflowStageType;
    isInitialStage?: boolean;
    isFinalStage?: boolean;
    inputDataSourceId?: number;
    targetDataSourceId?: number;
}

export interface UpdateWorkflowStageRequest {
    name?: string;
    description?: string;
    stageOrder?: number;
    stageType?: WorkflowStageType;
    isInitialStage?: boolean;
    isFinalStage?: boolean;
    inputDataSourceId?: number;
    targetDataSourceId?: number;
}

export interface CreateWorkflowStageConnectionRequest {
    fromStageId: number;
    toStageId: number;
    condition?: string;
}

export interface UpdateWorkflowStageConnectionRequest {
    fromStageId: number;
    toStageId: number;
    condition?: string;
}

export interface CreateWorkflowStageAssignmentRequest {
    workflowStageId: number;
    projectMemberId: number;
}

export interface WorkflowStageWithUsers extends WorkflowStage {
    assignedUsers: ProjectMember[];
}

// Utility functions for stage types
export const WorkflowStageTypeLabels: Record<WorkflowStageType, string> = {
    [WorkflowStageType.ANNOTATION]: "Annotation",
    [WorkflowStageType.SUSPENDED]: "Suspended",
    [WorkflowStageType.DEFERRED]: "Deferred",
    [WorkflowStageType.REVIEW]: "Review",
    [WorkflowStageType.REQUIRES_CHANGES]: "Requires Changes",
    [WorkflowStageType.ACCEPTED]: "Accepted"
};

export const formatStageType = (type: WorkflowStageType): string => {
    return WorkflowStageTypeLabels[type] || type;
};